File: tests/test_dataset_gallery.php
<?php
/* PHP twin of tests/test_dataset_gallery.py — asserts api/datasets.php agrees with
   dev_server.py on the per-dataset image gallery: what lands on disk, what the
   `file` field may look like (it becomes a URL in the viewer), and how a Save
   reconciles metadata.json against the gallery/ folder.

   The two backends must produce the SAME folder for the same input: an operator who
   moves a deployment from the Python dev server to a PHP host — or the other way —
   must find the same images with the same names and captions.

   datasets.php is loaded in library mode (LUMEN_DATASETS_LIB), so it defines its
   helpers without emitting headers, starting a session, or running the router.
     php tests/test_dataset_gallery.php                                          */
declare(strict_types=1);

define('LUMEN_DATASETS_LIB', true);
require_once __DIR__ . '/../api/datasets.php';

$root = sys_get_temp_dir() . '/lumen-gallery-' . bin2hex(random_bytes(4));
$ds   = $root . '/DATA_WEB/fixed/demo';
@mkdir($ds, 0777, true);

function rrm(string $d): void {
    if (!is_dir($d)) return;
    foreach (scandir($d) ?: [] as $f) {
        if ($f === '.' || $f === '..') continue;
        $p = "$d/$f";
        is_dir($p) ? rrm($p) : @unlink($p);
    }
    @rmdir($d);
}
register_shutdown_function(function () use ($root) { rrm($root); });

$fails = 0;
function check(string $name, $cond): void {
    global $fails;
    echo ($cond ? "  ok   " : "  FAIL ") . "$name\n";
    if (!$cond) $fails++;
}
function section(string $s): void { echo "\n$s\n"; }

$PNG  = hex2bin('89504e470d0a1a0a') . str_repeat("\x00", 40);
$JPEG = "\xff\xd8\xff" . str_repeat("\x00", 40);
$GIF  = 'GIF89a' . str_repeat("\x00", 40);
$WEBP = 'RIFF' . "\x00\x00\x00\x00" . 'WEBP' . str_repeat("\x00", 40);
$durl = fn(string $b) => 'data:image/png;base64,' . base64_encode($b);

$writeMeta = function (array $doc) use ($ds) {
    file_put_contents($ds . '/metadata.json', json_encode($doc));
};
$meta  = fn() => json_decode(file_get_contents($ds . '/metadata.json'), true);
$files = function () use ($ds) {
    $g = $ds . '/gallery';
    if (!is_dir($g)) return [];
    $f = array_values(array_diff(scandir($g), ['.', '..']));
    sort($f);
    return $f;
};
$reset = function () use ($ds, $writeMeta) {
    rrm($ds . '/gallery');
    $writeMeta(['name' => 'Demo', 'type' => 'fixed']);
};

// ── Upload ───────────────────────────────────────────────────────────────────
section('Upload');
$reset();
[$st, $pl] = gallery_add('fixed/demo', $ds, ['image' => $durl($PNG), 'filename' => 'shot.png', 'caption' => 'Vue sagittale']);
check('writes the file', $st === 200 && $files() === ['shot.png']);
check('records the entry', ($meta()['gallery'][0]['file'] ?? '') === 'shot.png');
check('keeps the caption', ($meta()['gallery'][0]['caption'] ?? '') === 'Vue sagittale');
check('returns a public url', ($pl['url'] ?? '') === 'DATA_WEB/fixed/demo/gallery/shot.png');

$reset();
$okAll = true;
foreach ([[$PNG, 'png'], [$JPEG, 'jpg'], [$GIF, 'gif'], [$WEBP, 'webp']] as [$raw, $ext]) {
    [$s, ] = gallery_add('fixed/demo', $ds, ['image' => $durl($raw), 'filename' => "f-$ext.bin"]);
    if ($s !== 200) $okAll = false;
}
check('accepts png/jpeg/gif/webp', $okAll && count($files()) === 4);

$reset();
[$st, $pl] = gallery_add('fixed/demo', $ds, ['image' => $durl($JPEG), 'filename' => 'liar.png']);
check('magic bytes decide the extension, not the name', ($pl['item']['file'] ?? '') === 'liar.jpg');

$reset();
[$st, ] = gallery_add('fixed/demo', $ds, ['image' => $durl('<?php system($_GET["c"]); ?>'), 'filename' => 'evil.png']);
check('non-image refused before any write', $st === 400 && $files() === [] && !isset($meta()['gallery']));

$reset();
[$st, ] = gallery_add('fixed/demo', $ds, ['image' => $durl($PNG . str_repeat("\x00", MAX_GALLERY_BYTES))]);
check('oversized payload refused', $st === 400 && $files() === []);

$reset();
[$st, ] = gallery_add('fixed/demo', $ds, ['image' => 'https://evil.example/x.png']);
check('non-data URL refused', $st === 400);

$reset();
gallery_add('fixed/demo', $ds, ['image' => $durl($PNG), 'filename' => 'Coupe Annotée !.png']);
gallery_add('fixed/demo', $ds, ['image' => $durl($PNG), 'filename' => 'Coupe Annotée !.png']);
check('name slugged and de-duplicated (same slug as Python)',
      $files() === ['coupe-annot-e-1.png', 'coupe-annot-e.png']);

$reset();
[$st, $pl] = gallery_add('fixed/demo', $ds, ['image' => $durl($PNG), 'filename' => '../../../../evil.png']);
check('traversal in the name cannot escape gallery/',
      $st === 200 && ($pl['item']['file'] ?? '') === 'evil.png'
      && $files() === ['evil.png'] && !is_file($root . '/evil.png'));

$reset();
for ($i = 0; $i < MAX_GALLERY_ITEMS; $i++) gallery_add('fixed/demo', $ds, ['image' => $durl($PNG), 'filename' => "i$i.png"]);
[$st, ] = gallery_add('fixed/demo', $ds, ['image' => $durl($PNG), 'filename' => 'overflow.png']);
check('count capped at ' . MAX_GALLERY_ITEMS, $st === 409 && count($files()) === MAX_GALLERY_ITEMS);

// Character count without relying on mbstring — this suite must run on a host that
// does not load it, which is exactly the configuration gallery_clean_text guards for.
$ulen = fn(string $s) => count(preg_split('//u', $s, -1, PREG_SPLIT_NO_EMPTY) ?: []);

$reset();
[, $pl] = gallery_add('fixed/demo', $ds, ['image' => $durl($PNG), 'caption' => "ligne 1\nligne 2\r\n" . str_repeat('x', 900)]);
$cap = $pl['item']['caption'] ?? '';
check('caption flattened and clamped to 400', strpos($cap, "\n") === false && $ulen($cap) === 400);

$reset();
// An accented caption cut on BYTES would produce invalid UTF-8 and json_encode would
// return false, blanking metadata.json. Clamp on characters, and prove the file survives.
[, $pl] = gallery_add('fixed/demo', $ds, ['image' => $durl($PNG), 'caption' => str_repeat('é', 900)]);
$cap = $pl['item']['caption'] ?? '';
check('accented caption clamped on characters, not bytes', $ulen($cap) === 400);
check('clamped caption is valid UTF-8', $cap === (string)@iconv('UTF-8', 'UTF-8//IGNORE', $cap));
check('metadata.json survives an accented caption',
      is_array($meta()) && ($meta()['gallery'][0]['caption'] ?? '') === $cap);

// ── Delete ───────────────────────────────────────────────────────────────────
section('Delete');
$reset();
gallery_add('fixed/demo', $ds, ['image' => $durl($PNG), 'filename' => 'a.png']);
gallery_add('fixed/demo', $ds, ['image' => $durl($PNG), 'filename' => 'b.png']);
[$st, $pl] = gallery_delete($ds, 'a.png');
check('removes file and entry', $st === 200 && $files() === ['b.png']
      && array_column($pl['gallery'], 'file') === ['b.png']);

$bad = true;
foreach (['../metadata.json', '../../fixed/demo/metadata.json', 'sub/a.png', 'a.php'] as $t) {
    [$s, ] = gallery_delete($ds, $t);
    if ($s !== 400) $bad = false;
}
check('traversal targets refused', $bad && is_file($ds . '/metadata.json'));

$reset();
gallery_add('fixed/demo', $ds, ['image' => $durl($PNG), 'filename' => 'a.png']);
[$st, ] = gallery_delete($ds, 'gallery/a.png');
check('"gallery/x.png" form accepted', $st === 200 && $files() === []);

// ── Reconciliation on save ───────────────────────────────────────────────────
section('Reconciliation on save');
$reset();
gallery_add('fixed/demo', $ds, ['image' => $durl($PNG), 'filename' => 'a.png', 'caption' => 'Vue A']);
gallery_add('fixed/demo', $ds, ['image' => $durl($PNG), 'filename' => 'b.png', 'caption' => 'Vue B']);
$stored = $meta()['gallery'];
$g = gallery_reconcile(['gallery' => [['file' => 'a.png', 'caption' => 'Vue A']]], $ds, $stored);
check('stale draft cannot drop a concurrent upload', array_column($g, 'file') === ['a.png', 'b.png']);
check('re-appended image keeps its stored caption', ($g[1]['caption'] ?? '') === 'Vue B');

$g = gallery_reconcile(['gallery' => [
    ['file' => 'a.png'], ['file' => 'ghost.png'],
    ['file' => '../../../etc/passwd'], ['file' => 'x.php'], 'not-an-object',
]], $ds, null);
check('entries with no file on disk are dropped', array_column($g, 'file') === ['a.png', 'b.png']);

$g = gallery_reconcile(['gallery' => [
    ['file' => 'b.png', 'caption' => 'second devient premier'], ['file' => 'a.png'],
]], $ds, null);
check('draft order and captions win', array_column($g, 'file') === ['b.png', 'a.png']
      && ($g[0]['caption'] ?? '') === 'second devient premier');

$g = gallery_reconcile(['gallery' => [['file' => 'a.png'], ['file' => 'a.png']]], $ds, null);
check('duplicates collapse', count(array_filter($g, fn($e) => $e['file'] === 'a.png')) === 1);

$reset();
check('no gallery folder yields an empty list', gallery_reconcile(['name' => 'Demo'], $ds, null) === []);

// ── Cross-backend agreement ──────────────────────────────────────────────────
// A folder written by the Python engine must be read identically here. The Python
// side produced these names/captions in tests/test_dataset_gallery.py.
section('Cross-backend');
$reset();
@mkdir($ds . '/gallery', 0777, true);
file_put_contents($ds . '/gallery/coupe-annot-e.png', $PNG);
file_put_contents($ds . '/gallery/liar.jpg', $JPEG);
$writeMeta(['name' => 'Demo', 'gallery' => [
    ['file' => 'coupe-annot-e.png', 'caption' => 'Vue sagittale', 'added' => '2026-08-08T12:00:00'],
    ['file' => 'liar.jpg'],
]]);
$g = gallery_reconcile($meta(), $ds, null);
check('reads a Python-written gallery unchanged',
      array_column($g, 'file') === ['coupe-annot-e.png', 'liar.jpg']
      && ($g[0]['caption'] ?? '') === 'Vue sagittale'
      && ($g[0]['added'] ?? '') === '2026-08-08T12:00:00');

echo "\n" . ($fails === 0 ? "ALL PASS\n" : "$fails FAILURE(S)\n");
exit($fails === 0 ? 0 : 1);

File: tests/test_dataset_gallery.py
"""Unit tests for the per-dataset image gallery (web v1.45.0).

Covers the storage engine in dev_server.py: what is accepted onto disk, what the
`file` field is allowed to look like (it becomes a URL in the viewer), and the
reconciliation that keeps metadata.json honest against the gallery/ folder.

Run: py tests/test_dataset_gallery.py
"""
import base64
import json
import os
import shutil
import sys
import tempfile
import unittest
from pathlib import Path

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
sys.path.insert(0, ROOT)
import dev_server  # noqa: E402

PNG = bytes.fromhex("89504e470d0a1a0a") + b"\x00" * 40
JPEG = b"\xff\xd8\xff" + b"\x00" * 40
GIF = b"GIF89a" + b"\x00" * 40
WEBP = b"RIFF" + b"\x00\x00\x00\x00" + b"WEBP" + b"\x00" * 40


def durl(raw: bytes, mime: str = "image/png") -> str:
    return f"data:{mime};base64," + base64.b64encode(raw).decode()


class GalleryCase(unittest.TestCase):
    def setUp(self):
        self.tmp = Path(tempfile.mkdtemp())
        self._orig_root = dev_server.DATA_WEB
        dev_server.DATA_WEB = self.tmp
        self.ds_dir = self.tmp / "3d" / "demo"
        self.ds_dir.mkdir(parents=True)
        self._write_meta({"name": "Demo", "type": "3d"})

    def tearDown(self):
        dev_server.DATA_WEB = self._orig_root
        shutil.rmtree(self.tmp, ignore_errors=True)

    # helpers
    def _write_meta(self, doc):
        (self.ds_dir / "metadata.json").write_text(json.dumps(doc), encoding="utf-8")

    def _meta(self):
        return json.loads((self.ds_dir / "metadata.json").read_text(encoding="utf-8"))

    def _add(self, raw=PNG, name="shot.png", **kw):
        body = {"image": durl(raw), "filename": name}
        body.update(kw)
        return dev_server._gallery_add("3d/demo", body)

    def _files(self):
        gdir = self.ds_dir / "gallery"
        return sorted(p.name for p in gdir.iterdir()) if gdir.is_dir() else []


class TestUpload(GalleryCase):
    def test_writes_file_and_records_entry(self):
        status, payload = self._add(caption="Vue sagittale")
        self.assertEqual(status, 200)
        self.assertEqual(self._files(), ["shot.png"])
        entry = self._meta()["gallery"][0]
        self.assertEqual(entry["file"], "shot.png")
        self.assertEqual(entry["caption"], "Vue sagittale")
        self.assertEqual(payload["url"], "DATA_WEB/3d/demo/gallery/shot.png")

    def test_every_supported_format_is_accepted(self):
        for raw, ext in ((PNG, "png"), (JPEG, "jpg"), (GIF, "gif"), (WEBP, "webp")):
            status, _ = self._add(raw, name=f"f-{ext}.bin")
            self.assertEqual(status, 200, ext)
        self.assertEqual(len(self._files()), 4)

    def test_extension_comes_from_magic_bytes_not_the_filename(self):
        # A JPEG announced as .png must land as .jpg: the name never decides the type.
        status, payload = self._add(JPEG, name="liar.png")
        self.assertEqual(status, 200)
        self.assertEqual(payload["item"]["file"], "liar.jpg")

    def test_non_image_is_refused_before_any_write(self):
        status, _ = self._add(b"<?php system($_GET['c']); ?>", name="evil.png")
        self.assertEqual(status, 400)
        self.assertEqual(self._files(), [])
        self.assertNotIn("gallery", self._meta())

    def test_oversized_payload_is_refused(self):
        status, _ = self._add(PNG + b"\x00" * dev_server.MAX_GALLERY_BYTES)
        self.assertEqual(status, 400)
        self.assertEqual(self._files(), [])

    def test_non_data_url_is_refused(self):
        status, _ = dev_server._gallery_add("3d/demo", {"image": "https://evil.example/x.png"})
        self.assertEqual(status, 400)

    def test_filename_is_slugged_and_deduplicated(self):
        self._add(name="Coupe Annotée !.png")
        self._add(name="Coupe Annotée !.png")
        self.assertEqual(self._files(), ["coupe-annot-e-1.png", "coupe-annot-e.png"])

    def test_traversal_in_filename_cannot_escape_the_gallery_folder(self):
        status, payload = self._add(name="../../../../evil.png")
        self.assertEqual(status, 200)
        self.assertEqual(payload["item"]["file"], "evil.png")   # slugged to a bare name
        self.assertEqual(self._files(), ["evil.png"])
        self.assertFalse((self.tmp / "evil.png").exists())

    def test_bad_dataset_id_is_refused(self):
        for bad in ("../../api", "3d", "2d", "secrets/keys", "3d/../../api", "fixed/demo"):
            status, _ = dev_server._gallery_add(bad, {"image": durl(PNG)})
            self.assertEqual(status, 400, bad)

    def test_count_is_capped(self):
        for i in range(dev_server.MAX_GALLERY_ITEMS):
            self.assertEqual(self._add(name=f"i{i}.png")[0], 200)
        status, _ = self._add(name="overflow.png")
        self.assertEqual(status, 409)
        self.assertEqual(len(self._files()), dev_server.MAX_GALLERY_ITEMS)

    def test_caption_is_flattened_and_clamped(self):
        _, payload = self._add(caption="ligne 1\nligne 2\r\n" + "x" * 900)
        caption = payload["item"]["caption"]
        self.assertNotIn("\n", caption)
        self.assertEqual(len(caption), 400)


class TestDelete(GalleryCase):
    def test_removes_file_and_entry(self):
        self._add(name="a.png")
        self._add(name="b.png")
        status, payload = dev_server._gallery_delete("3d/demo", "a.png")
        self.assertEqual(status, 200)
        self.assertEqual(self._files(), ["b.png"])
        self.assertEqual([e["file"] for e in payload["gallery"]], ["b.png"])

    def test_traversal_target_is_refused(self):
        self._add(name="a.png")
        for bad in ("../metadata.json", "../../3d/demo/metadata.json", "sub/a.png", "a.php"):
            status, _ = dev_server._gallery_delete("3d/demo", bad)
            self.assertEqual(status, 400, bad)
        self.assertTrue((self.ds_dir / "metadata.json").exists())

    def test_gallery_prefix_form_is_accepted(self):
        # The viewer stores bare names, but a hand-written "gallery/x.png" still resolves.
        self._add(name="a.png")
        status, _ = dev_server._gallery_delete("3d/demo", "gallery/a.png")
        self.assertEqual(status, 200)
        self.assertEqual(self._files(), [])


class TestReconcileOnSave(GalleryCase):
    """metadata.json's gallery decides ORDER and CAPTIONS; the folder decides WHICH."""

    def test_stale_draft_cannot_drop_a_concurrent_upload(self):
        self._add(name="a.png", caption="Vue A")
        self._add(name="b.png", caption="Vue B")
        # A form opened before b.png existed saves back only a.png.
        dev_server._save_dataset("3d/demo", {"name": "Demo", "gallery": [{"file": "a.png", "caption": "Vue A"}]})
        gallery = self._meta()["gallery"]
        self.assertEqual([e["file"] for e in gallery], ["a.png", "b.png"])
        # …and b.png keeps the caption it was uploaded with.
        self.assertEqual(gallery[1]["caption"], "Vue B")

    def test_entries_without_a_file_on_disk_are_dropped(self):
        self._add(name="a.png")
        dev_server._save_dataset("3d/demo", {"gallery": [
            {"file": "a.png"},
            {"file": "ghost.png"},
            {"file": "../../../etc/passwd"},
            {"file": "x.php"},
            "not-an-object",
        ]})
        self.assertEqual([e["file"] for e in self._meta()["gallery"]], ["a.png"])

    def test_order_and_captions_from_the_draft_are_kept(self):
        self._add(name="a.png")
        self._add(name="b.png")
        dev_server._save_dataset("3d/demo", {"gallery": [
            {"file": "b.png", "caption": "second devient premier"},
            {"file": "a.png"},
        ]})
        gallery = self._meta()["gallery"]
        self.assertEqual([e["file"] for e in gallery], ["b.png", "a.png"])
        self.assertEqual(gallery[0]["caption"], "second devient premier")

    def test_duplicate_entries_collapse(self):
        self._add(name="a.png")
        dev_server._save_dataset("3d/demo", {"gallery": [{"file": "a.png"}, {"file": "a.png"}]})
        self.assertEqual(len(self._meta()["gallery"]), 1)

    def test_key_is_removed_when_no_image_remains(self):
        self._add(name="a.png")
        dev_server._gallery_delete("3d/demo", "a.png")
        dev_server._save_dataset("3d/demo", {"name": "Demo"})
        self.assertNotIn("gallery", self._meta())

    def test_a_dataset_without_a_gallery_folder_is_unaffected(self):
        dev_server._save_dataset("3d/demo", {"name": "Demo"})
        self.assertNotIn("gallery", self._meta())


class TestCatalogExposure(GalleryCase):
    def test_gallery_reaches_the_public_catalog(self):
        self._write_meta({"name": "Demo", "type": "3d", "configured": True,
                          "dimensions": {"x": 4, "y": 4, "z": 4, "c": 1},
                          "channels": [{"name": "c0"}]})
        self._add(name="a.png", caption="Vue A")
        dev_server._CATALOG_CACHE["sig"] = None
        entry = next(d for d in dev_server._build_catalog() if d["id"] == "3d/demo")
        self.assertEqual([e["file"] for e in entry["gallery"]], ["a.png"])
        self.assertEqual(entry["gallery"][0]["caption"], "Vue A")


if __name__ == "__main__":
    unittest.main(verbosity=2)
